# IxodoideaDB-annotation-pipeline

### 01 Prerequisites
#### 1.1 Software
+ BUSCO (https://busco.ezlab.org/) (compleasm(https://github.com/huangnengCSU/compleasm), compleasm: a faster and more accurate reimplementation of BUSCO)
+ HiTE (https://github.com/CSU-KangHu/HiTE)
+ Bakta (https://github.com/oschwengers/bakta)
+ geNomad (https://github.com/apcamargo/genomad)


#### 1.2 Database
+ arachnida_odb10 (https://busco-data.ezlab.org/v5/data/lineages/arachnida_odb10.2024-01-08.tar.gz)
+ db-light (https://zenodo.org/record/7669534/files/db-light.tar.gz?download=1)
+ genomad_db (https://portal.nersc.gov/genomad/__data__/genomad_db_v1.9.tar.gz)
+ checkm2_database (https://zenodo.org/api/files/fd3bc532-cd84-4907-b078-2e05a1e46803/checkm2_database.tar.gz)


### 02 Genome assessment
#### 2.1 BUSCO v5.6.0
- genome.fna
- arachnida_odb10
```shell
busco -i ./tickdb/genome.fna -l ./arachnida_odb10 -o ./busco_output -m genome --cpu 12 --offline
```
```bash
cat ./output_dir/short_summary.specific.arachnida_odb10.out.txt
```
or you can try compleasm, a faster and more accurate reimplementation of BUSCO
#### 2.2 compleasm
```bash
python compleasm.py run -t 16 -l arachnida -L ./arachnida_odb10 -a genome.fna -o compleasm_output
```
Note: the organization of the lineage file downloaded by compleasm is different from that of BUSCO.

#### 2.3 CheckM2
- bacteriadb/
- checkm2_database
```bash
checkm2 predict --threads 16 --input ./bacteriadb --database_path ./CheckM2_database/uniref100.KO.1.dmnd --output_directory ./checkm2_output
```
------

### 3. Repeat annotation and genome mask
#### 3.1 EDTA 
- genome.fna
You got a genome and you want to get a high-quality TE annotation:
```bash
perl EDTA.pl --genome ../tickdb/genome.fna --step all --overwrite 1 -t 12 --sensitive 1 --anno 1 --force 1 evaluate 1
```
```bash
# Using EDTA to convert hard masking into soft masking
perl ./util/make_masked.pl -genome genome.fna -minlen 80 -hardmask 0 -t 10 -rmout genome.fna.mod.EDTA.TEanno.out
```
#### 3.2 HiTE
- genome.fna
Use the TE library generated by HiTE to annotate the genome. This will produce annotation files such as `HiTE.out`, `HiTE.gff`, and `HiTE.tbl`. 
```bash
python main.py --genome ../tickdb/genome.fna --outdir ../hite_output --thread 12 --annotate 1 --plant 0
```
HiTE outputs many temporary files, which allow you to quickly restore the previous 
running state (use `--recover 1`) in case of any interruption during the running process. If
the pipeline completes successfully, the output directory should look like the following:
```shell
output_dir/
├── longest_repeats_*.fa
├── longest_repeats_*.flanked.fa
├── confident_tir_*.fa
├── confident_helitron_*.fa
├── confident_non_ltr_*.fa
├── confident_other_*.fa
├── confident_ltr_cut.fa.cons
├── confident_TE.cons.fa
├── HiTE.out (require `--annotate 1`)
├── HiTE.gff (require `--annotate 1`)
└── HiTE.tbl (require `--annotate 1`)
```
------

### 04 Structural annotation of the tick genome by EGAPx
------
### 05 Standardized annotation of Tick-Borne Bacterial Genomes by Bakta
- genome.fna
- db-light

If required, or desired, the AMRfinderPlus-DB can also be updated manually: 
```bash
mkdir -p db-light/amrfinderplus-db
```
```bash
amrfinder_update --force_update --database db-light/amrfinderplus-db/
```
```bash
bakta --db ./db-light --output ./bakta_output ./bacteriadb/genome.fna
```
It provides dbxref-rich, sORF-including and taxon-independent annotations in machine-readable JSON & bioinformatics standard file formats for automated downstream analysis.
------

### 06 Automated annotation of Tick-Borne Virus Genomes by geNomad
- genome.fna
- genomad_db

```bash
genomad end-to-end --cleanup --splits 8 ./virusdb/genome.fna.gz ./genomad_output ./genomad_db
